File: dsa-csharp/solutions/ArrayAndHashing/Medium/GroupAnagrams.cs
// 49. Group Anagrams
// https://leetcode.com/problems/group-anagrams/
//
// Given an array of strings strs, group the anagrams together. You can return
// the answer in any order. An anagram is a word or phrase formed by rearranging
// the letters of a different word or phrase, typically using all the original
// letters exactly once.
//
// APPROACH: Sorted string as HashMap key
//
// KEY INSIGHT: All anagrams sort to the same string.
// Use the sorted form as a unique key to group anagrams together.
//
// HOW IT WORKS:
// - For each string, sort its characters to get the canonical form.
// - Use sorted string as dictionary key.
// - Append the original string to its group.
//
// ALT APPROACH: Frequency array as key — O(n*k) but messier in C#.
// Sorting is O(n * k log k) but cleaner and fast enough.
//
// TIME: O(n * k log k) — n strings, k = average string length
// SPACE: O(n * k) — storing all strings in the map

public class GroupAnagrams
{
    public IList<IList<string>> GroupAnagramsSolution(string[] strs)
    {
        Dictionary<string, List<string>> groups = new();  // Maps sorted key → anagram group

        foreach(string s in strs)
        {
            char[] chars = s.ToCharArray();
            Array.Sort(chars);                             // Sort chars to get canonical form
            var key = new string(chars);                   // "eat" → "aet"

            if(!groups.ContainsKey(key))
                groups[key] = new List<string>();          // New group
            groups[key].Add(s);                            // Add string to its anagram group
        }

        List<IList<string>> result = new();

        foreach(var values in groups.Values) result.Add(values);  // Collect all groups

        return result;

    }
}

File: dsa-python/solutions/array_and_hashing/medium/group_anagrams.py
# 49. Group Anagrams
# https://leetcode.com/problems/group-anagrams/
#
# Given an array of strings strs, group the anagrams together. You can return
# the answer in any order. An anagram is a word or phrase formed by rearranging
# the letters of a different word or phrase, typically using all the original
# letters exactly once.
#
# APPROACH: Frequency tuple as HashMap key
#
# KEY INSIGHT: All anagrams have identical character frequencies.
# Use a tuple of 26 character counts as the unique key for each group.
#
# HOW IT WORKS:
# - For each string, build a frequency array (26 slots for a-z).
# - Convert to tuple (hashable) and use as dictionary key.
# - Append the original string to its group.
#
# ALT APPROACH: Sort each string as key — simpler but O(n * k log k).
# Frequency tuple is O(n * k) since no sorting needed.
#
# TIME: O(n * k) — n strings, k = average string length
# SPACE: O(n * k) — storing all strings in the map


class GroupAnagrams:
    def group_anagrams(self, strs: list[str]) -> list[list[str]]:
        groups = {}  # Maps frequency tuple → list of anagrams

        for s in strs:
            count = [0] * 26                       # Frequency array for a-z
            for c in s:
                count[ord(c) - ord('a')] += 1      # Count each character
            key = tuple(count)                     # Convert to hashable key
            if key not in groups:
                groups[key] = []                   # New group
            groups[key].append(s)                  # Add string to its anagram group

        return list(groups.values())  # Return all groups
